packages/agreement-process/src/services/agreementService.ts

```diff
@@ -277,3 +346,160 @@ export async function updateAgreementLogic({
     agreementToBeUpdated.metadata.version
   );
 }
+
+// eslint-disable-next-line sonarjs/cognitive-complexity
+export async function upgradeAgreementLogic({
+  agreementId,
+  authData,
+  agreementQuery,
+  eserviceQuery,
+  tenantQuery,
+  fileCopy,
+}: {
+  agreementId: string;
+  authData: AuthData;
+  agreementQuery: AgreementQuery;
+  eserviceQuery: EserviceQuery;
+  tenantQuery: TenantQuery;
+  fileCopy: (
+    container: string,
+    sourcePath: string,
+    destinationPath: string,
+    destinationFileName: string,
+    docName: string
+  ) => Promise<string>;
+}): Promise<Array<CreateEvent<AgreementEvent>>> {
+  const agreementToBeUpgraded = await agreementQuery.getAgreementById(
+    agreementId
+  );
+  const tenant = await tenantQuery.getTenantById(authData.organizationId);
+  assertTenantExist(authData.organizationId, tenant);
+  assertAgreementExist(agreementId, agreementToBeUpgraded);
+  assertRequesterIsConsumer(
+    agreementToBeUpgraded.data.consumerId,
+    authData.organizationId
+  );
+
+  const upgradableStates: AgreementState[] = [
+    agreementState.active,
+    agreementState.suspended,
+  ];
+
+  assertExpectedState(
+    agreementId,
+    agreementToBeUpgraded.data.state,
+    upgradableStates
+  );
+
+  const eservice = await eserviceQuery.getEServiceById(
+    agreementToBeUpgraded.data.eserviceId
+  );
+  assertEServiceExist(agreementToBeUpgraded.data.eserviceId, eservice);
+
+  const newDescriptor = eservice.data.descriptors.find(
+    (d) => d.state === descriptorState.published
+  );
+  if (newDescriptor === undefined) {
+    throw publishedDescriptorNotFound(agreementToBeUpgraded.data.eserviceId);
+  }
+  const latestDescriptorVersion = z
+    .preprocess((x) => Number(x), z.number())
+    .safeParse(newDescriptor.version);
+  if (!latestDescriptorVersion.success) {
+    throw unexpectedVersionFormat(eservice.data.id, newDescriptor.id);
+  }
+
+  const currentDescriptor = eservice.data.descriptors.find(
+    (d) => d.id === agreementToBeUpgraded.data.descriptorId
+  );
+  if (currentDescriptor === undefined) {
+    throw descriptorNotFound(
+      eservice.data.id,
+      agreementToBeUpgraded.data.descriptorId
+    );
+  }
+
+  const currentVersion = z
+    .preprocess((x) => Number(x), z.number())
+    .safeParse(currentDescriptor.version);
+  if (!currentVersion.success) {
+    throw unexpectedVersionFormat(eservice.data.id, currentDescriptor.id);
+  }
+
+  if (latestDescriptorVersion.data <= currentVersion.data) {
+    throw noNewerDescriptor(eservice.data.id, currentDescriptor.id);
+  }
+
+  validateCertifiedAttributes(newDescriptor, tenant.data);
```

Comment:
inerent to [this check](https://github.com/pagopa/interop-be-agreement-process/blob/66781549a6db2470d8c407965b7561d1fe493107/src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala#L250) could be missing condition   ( .whenA(eService.producerId != requesterOrgId) ) ??
if necessary we can use validation `assertRequesterIsConsumer` method  (meanwhile I wrote this comment I think that the method's name is wrong 😅 )